### ISSUE TITLE
add a MaxBoundEvent to the TimeConstraint class

### ISSUE THREAD

**Issue by theod** [OSSIA/API#22]
I keep this proposition as a question as it need to be discussed together.
it allows to create alternative branch depending if the interaction happen or not before the max bound.
the idea have been fully described on the dev list.

**Comment by nvuaille** [comment]
I realize yesterday that one can use 'fake' osc device as an internal memory to store global var. So I was able to create a scenario resolving the blue yeti problem (a player run and press a button at the end. A short delay means cheat, a too long one means player left) with a system of flag.

That's cool but not very user friendly, I agree. 
But I think the API should stay minimalist so we shouldn't add an element that doesn't allow anything new.

**Comment by theod** [comment]
yes the solution you proposed was quite the same than Blue Yeti proposed in the Runner example (that you can read here : https://www.dropbox.com/s/8prlpioqyg750y2/M_%20use%20case%20Arena.pdf?dl=0 ). in this solution Blue Yeti resolved this case by comparing the elapsed time of the constraint and its max duration bound (considering i-score elements as part of a local device).

I understand your reluctance to add a new element to the formalism and this always have to be done cautiously. however the solution we discussed with @bltzr, @jcelerier and @blueyeti to write the JIM paper have the great advantage to propose a very readable solution that avoid users to look into a local device to understand what is going on.

IHMO that is typically the kind of problem that i-score have to solve ingeniously if we want to provide a tool that makes people's life easier when they deal with those kinds of logics-temporal authoring issue.

I'll forward you the article in a private email.

**Comment by nvuaille** [comment]
> IHMO that is typically the kind of problem that i-score have to solve ingeniously

I agree with you. You talk here about i-score, not about the model. I think that we can do an interface in i-score to make it simpler and clear and I think that someone able to use directly the model will be able to do the trick with internal device.
So we need to modify the software, not the model.

**Comment by jcelerier** [comment]
> I understand your reluctance to add a new element to the formalism and this always have to be done cautiously.

I personnally have a main reluctance (at least for doing this today) :
The software almost works. There are still important issues before releasing anything as final, but in terms of useability, I could get students (both french, and spanish at Bilbao university) to write their own scores with logic in about one hour of class and 15 minutes of personal explanation. So there is no difficulty problem to use the model, and it is already (I think) much easier to write something temporal in i-score than in other environments.

 The JIM article mostly is about providing higher-level features like C++ offers higher-level features on the same underlying model (assembly) than C. History even gives us a hint at what may happen : the first C++ compiler actually compiled to C, which then was compiled to assembly. Now, let's call the description in the JIM paper the "JIM" formalism for the sake of this discussion; it is clear that today the way to go would be to have the "JIM" formalism in a JIM-score software which then calls the OSSIA API with the OSSIA primitives. But maybe in five years when the software has some hard bugs or performance issue due to small mismatches between the two, or the need for converting, somebody will come up and say "hey look at this old OSSIA API to which we have to translate everything that we do in JIM-score, why shouldn't we do a JIM API that directly implements the requested features ?", and the circle of the ANR projects will go round once again).

Also, doing such a change would necessarily break stuff (because we are not perfect programmers). And for instance, Georges is currently doing work with i-score, @blueyeti would like to engage working with i-score soon, I heard that Pascal was about to start doing serious work with i-score too, and there are people going to the website and downloading it.
Doing this change would break i-score for *at least* a month, with potential repercussions further down the line, and most importantly this will cause the projects of all these users to be delayed further, which will ultimately cause them to do their work in another environment that i-score.

While I understand the need of having a "perfect" model that solves all the use case brillantly, this is not going to happen, at least not outside the realm of conference papers, due to the lack of manpower. 
Besides, here is some relevant reading : https://www.jwz.org/doc/worse-is-better.html
Basically, C and UNIX sucks conceptually with regards to Common Lisp; however, aiming for more simplicity on the developer's side allowed them to take over the computing world. So, question : what is the goal of i-score (the software) : trying to be the most simple, conceptually perfect environment, or being something that is **actually used in production**, which, in my opinion, could be in some months if the "central" parts of the software do not change ?

**Comment by theod** [comment]
yes yes I fully agree this is not urgent !
that's why I've put the https://github.com/OSSIA/i-score/issues/142 issue on release/1.1 (and don't hesitate to postpone it for later). I was just opening a thread to not forget this proposition. I'll won't disturb you anymore from the main goal which is having i-score working for production !
on my side I'll work on this on a separate branch to check the consequence of this idea then I'll share you those results once I'll have interesting results.

**Comment by cbossut** [comment]
In the way I thought it, it is already possible in the formalism. You
should be able to test the actual duration with the max duration in an
expression, therefore add this case in a condition. Or you could even have
a boolean variable for the way the condition has been triggered, expression
or max duration.

2016-02-05 15:02 GMT+01:00 theod <notifications@github.com>:

> yes yes I fully agree this is not urgent !
> that's why I've put the OSSIA/i-score#142
> <https://github.com/OSSIA/i-score/issues/142> issue on release/1.1 (and
> don't hesitate to postpone it for later). I was just opening a thread to
> not forget this proposition. I'll won't disturb you anymore from the main
> goal which is having i-score working for production !
> on my side I'll work on this on a separate branch to check the consequence
> of this idea then I'll share you those results once I'll have interesting
> results.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/OSSIA/API/issues/22#issuecomment-180371350>.
>



-- 
Clément Bossut

**Comment by theod** [comment]
@cbossut sorry I didn't answer to your comment. 
As I said to @nvuaille : 
> the solution you proposed was quite the same than Blue Yeti proposed in the Runner example (that you can read here : https://www.dropbox.com/s/8prlpioqyg750y2/M_%20use%20case%20Arena.pdf?dl=0 ). in this solution Blue Yeti resolved this case by comparing the elapsed time of the constraint and its max duration bound (considering i-score elements as part of a local device).

the (maybe only visual) "Max bound event" solution I proposed is to make the representation of this as clear as possible because as you can see in the Arena example it is very ugly and complexe to write a such simple case.

then the discussion is more : should we introduce this as a new API TimeConstraint feature (which is my opinion as it seems not a big deal to add it and make it works) or should we keep the API as it is and make the complexe stuff with local device and so on behind the scene in i-score (which is the opinion of @jcelerier and @nvuaille) ?

then notice this proposition will also add some complexity to the CSP.

**Comment by nvuaille** [comment]
Not sure that CSP will be more complexe if this stuff is added to the API : a consistent scenario without  "Max bound event" will keep be consistent with  "Max bound event", as no duration will be changed, no ?

**Comment by theod** [comment]
I guess yes (I was not sure because I don't know how the CSP works actually).

**Comment by theod** [comment]
I've renamed the ticket because Pascal proposes a new feature that would make the MaxBoundEvent not optional.

**Comment by cbossut** [comment]
In my opinion, the API should be minimal complete, and adding this to the
API is not minimal.
This is said without considering anything else, so I understand that it
could not be the best solution.

2016-02-10 14:51 GMT+01:00 theod <notifications@github.com>:

> I guess yes (I was not sure because I don't know how the CSP works
> actually).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/OSSIA/API/issues/22#issuecomment-182384809>.
>



-- 
Clément Bossut

**Comment by theod** [comment]
as I said to clément here (https://github.com/OSSIA/API/issues/25#issuecomment-182978732)
I've finally understood what you mean @nvuaille and @jcelerier so sorry for this noise.
so I will give up to add this ugly feature ! 

but I'll add a new one ;-) to think about how to refer to time elements properties into an expression.
